[Feat] generalize step length

observingClouds opened the pull request:
## Describe your changes

This PR generalizes the time stepping to all datetime units.

Based on the temporal resolution of the datastore, the largest temporal unit that is able to represent the temporal resolution as integer is chosen.

If for example the temporal resolution is 120 min, the largest temporal unit being able to represent this is `hours`. Similarly, if the resolution is 119min, the unit would be `minutes`.

This will enable to train and infer on sub-hourly data consistently.

## Issue Link

eventually solves #171 

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [ ] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [x] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [x] the code is readable
- [x] the code is well tested
- [x] the code is documented (including return types and parameters)
- [x] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug

## Checklist for assignee

- [x] PR is up to date with the base branch
- [x] the tests pass
- [x] author has added an entry to the changelog (and designated the change as *added*, *changed* or *fixed*)
- Once the PR is ready to be merged, squash commits and merge the PR.

observingClouds committed:
WIP: easy changes

observingClouds committed:
WIP: largest int step unit

observingClouds committed:
WIP: subsampling

observingClouds committed:
remove unused import

observingClouds committed:
use python3.9 compatible typing

observingClouds committed:
set step_length with new type in test

observingClouds commented on neural_lam/datastore/npyfilesmeps/compute_standardization_stats.py:
```diff
@@ -299,7 +300,8 @@ def main(
         num_workers=n_workers,
         sampler=sampler_standard,
     )
-    used_subsample_len = (65 // step_length) * step_length
```
@leifdenby, @joeloskarsson, @sadamov is this a relict of Oskarsson et al. (2023) where you potentially had a 65 hour forecast (copilot is suggesting that)?

I do not quite get what this is supposed to do.

joeloskarsson replied:
Yes exactly, the MEPS data (that is a specific dataset that this datastore is written for) has 65 time steps in each sample that is loaded from disk.

observingClouds replied:
Thanks for the clarification. So in this particular case step_length has to be a multiple of 1 hour and cannot be e.g. 30min or so, right?

joeloskarsson reviewed (changes requested):
This is great! Having the step length as an int was really a quite ugly setup.

I left some small comments, but overall I agree with all the changes and the design.

observingClouds committed:
run docstring tests

observingClouds commented on neural_lam/utils.py:
```diff
@@ -408,3 +409,45 @@ def inverse_sigmoid(x):
     return torch.log(x_clamped / (1 - x_clamped))
+
+
+def get_integer_time(tdelta) -> Optional[tuple[int, str]]:
```
I added the `--doctest-modules` flag to pytest, so now the docstring examples given in this function will be tested without the need of additional test code.

observingClouds committed:
fix doctest

observingClouds committed:
document changes

observingClouds committed:
return 1, unknown as fallback

observingClouds committed:
meps step_length in hours

observingClouds committed:
add test

observingClouds committed:
fix abstract base class

observingClouds committed:
fix attr

observingClouds committed:
add plotting test

observingClouds committed:
ensure meps complient time unit

observingClouds committed:
simplify title formatting

observingClouds committed:
Merge branch 'main' into feat/generalize_step_length

joeloskarsson reviewed (changes requested):
Really great with the plotting test as well! I had one question, otherwise all looks good.

observingClouds commented on tests/test_plotting.py:
```diff
@@ -0,0 +1,185 @@
+    )
+
+    # Set time_step_int_unit (appears to be a typo in ar_model.py line 481)
+    # Should probably be time_step_unit based on line 159
```
I fixed this, but need to remove this comment

observingClouds committed:
Fix time step unit assignment in test_plotting.py

Removed incorrect assignment of time_step_int_unit and updated to use time_step_unit.

observingClouds commented on tests/test_plotting.py:
```diff
@@ -0,0 +1,181 @@
+
+    # Override time step information to test different scenarios
+    model.time_step_int = time_step
+    model.time_step_unit = time_unit
```
Looking at this again, I like to implement the test differently so that it is ensured that the timestep and its unit is correctly inferred from the dataset.

observingClouds committed:
include timestep inference in test suite

observingClouds committed:
improve comment

joeloskarsson reviewed (approved):
LGTM now! These tests are really great :)
